File: tasks-core-jvm2/src/main/java/org/funfix/tasks/Outcome.java
package org.funfix.tasks;

import org.jspecify.annotations.NullMarked;

import java.util.concurrent.ExecutionException;

/**
 * Represents the outcome of a finished task.
 *
 * @param <T> is the type of the value that the task completed with
 */
@NullMarked
public sealed interface Outcome<T>
    permits Outcome.Succeeded, Outcome.Failed, Outcome.Cancelled {

    /**
     * Returns the value of the task if it was successful, or throws an exception.
     *
     * @return the successful value (in case the outcome is a {@link Succeeded})
     *
     * @throws ExecutionException if the task failed with an exception
     * @throws CancellationException if the task was cancelled
     */
    T getOrThrow() throws ExecutionException, CancellationException;

    /**
     * Signals a successful result of the task.
     *
     * @param value is the value that the task completed with
     */
    record Succeeded<T>(T value) implements Outcome<T> {
        @Override
        public T getOrThrow() {
            return value;
        }
    }

    /**
     * Signals that the task failed.
     *
     * @param exception is the exception that the task failed with
     */
    record Failed<T>(Throwable exception) implements Outcome<T> {
        @Override
        public T getOrThrow() throws ExecutionException {
            throw new ExecutionException(exception);
        }
    }

    /**
     * Signals that the task was cancelled.
     */
    record Cancelled<T>() implements Outcome<T> {
        @Override
        public T getOrThrow() throws CancellationException {
            throw new CancellationException();
        }
    }

    /**
     * @return a {@link Succeeded} outcome.
     */
    static <T> Outcome<T> succeeded(final T value) {
        return new Succeeded<>(value);
    }

    /**
     * @return a {@link Failed} outcome.
     */
    static <T> Outcome<T> failed(final Throwable error) {
        return new Failed<>(error);
    }

    /**
     * @return a {@link Cancelled} outcome.
     */
    static <T> Outcome<T> cancelled() {
        return new Cancelled<>();
    }
}

File: tasks-core/src/jvmMain/kotlin/org/funfix/tasks/Outcome.kt
@file:Suppress("DataClassPrivateConstructor")

package org.funfix.tasks

import java.util.concurrent.ExecutionException

/**
 * Represents the outcome of a concurrent computation (e.g., `Fiber`).
 */
sealed interface Outcome<out T> {
    /**
     * Returns the value of the task if it was successful, or throws an exception.
     *
     * @return the successful value (in case the outcome is [Succeeded])
     *
     * @throws ExecutionException if the task failed with an exception ([Failed])
     * @throws CancellationException if the task was cancelled ([Cancelled])
     */
    @Throws(ExecutionException::class, CancellationException::class)
    fun getOrThrow(): T

    /**
     * Value signalling the concurrent job completed successfully with a result.
     *
     * To get an instance of this class, use [Outcome.succeeded]:
     * ```java
     * Outcome<String> outcome = Outcome.succeeded("Hello");
     * // ... alternatively:
     * Outcome<String> outcome = Outcome.Succeeded.instance("Hello");
     * ```
     *
     * Starting with Java 16 this can be used in pattern-matching:
     * ```java
     * if (outcome instanceof Outcome.Succeeded<String> succeeded) {
     *   System.out.println("Value is: " + succeeded.value());
     * }
     * ```
     */
    data class Succeeded<out T> private constructor(
        private val _value: T
    ) : Outcome<T> {
        val value: T
            @JvmName("value")
            get() = _value

        override fun getOrThrow(): T {
            return value
        }

        companion object {
            @JvmStatic
            @JvmName("instance")
            operator fun <T> invoke(value: T): Outcome<T> =
                Succeeded(value)
        }
    }

    /**
     * Value signaling that the concurrent job failed with an `exception`.
     *
     * To get an instance of this class, use [Outcome.failed]:
     * ```java
     * Outcome<String> outcome = Outcome.failed(new RuntimeException("Boom!"));
     * // ... alternatively:
     * Outcome<String> outcome = Outcome.Failed.instance(new RuntimeException("Boom!"));
     * ```
     *
     * Starting with Java 16 this can be used in pattern-matching:
     * ```java
     * if (outcome instanceof Outcome.Failed failed) {
     *    failed.exception().printStackTrace();
     * }
     * ```
     */
    data class Failed private constructor(
        private val _exception: Throwable
    ) : Outcome<Nothing> {
        val exception: Throwable
            @JvmName("exception")
            get() = _exception

        @Throws(ExecutionException::class)
        override fun getOrThrow(): Nothing {
            throw ExecutionException(exception)
        }

        companion object {
            @JvmStatic
            @JvmName("instance")
            operator fun <T> invoke(exception: Throwable): Outcome<T> =
                Failed(exception)
        }
    }

    /**
     * Value signalling that the concurrent job was cancelled.
     *
     * To get an instance of this class, in Java, use [Outcome.cancelled]:
     * ```java
     * Outcome<String> outcome = Outcome.cancelled()
     * // ... alternatively:
     * Outcome<String> outcome = Outcome.Cancelled.instance();
     * ```
     *
     * Starting with Java 16 this can be used in pattern-matching:
     * ```java
     * if (outcome instanceof Outcome.Cancelled cancelled) {
     *    // *cancelled* is the instance of type *Outcome.Cancelled*
     * }
     * ```
     */
    data object Cancelled: Outcome<Nothing> {
        @Throws(CancellationException::class)
        override fun getOrThrow(): Nothing {
            throw CancellationException()
        }

        @JvmStatic
        @JvmName("instance")
        operator fun <T> invoke(): Outcome<T> =
            Cancelled
    }

    companion object {
        @JvmStatic fun <T> succeeded(value: T): Outcome<T> =
            Succeeded(value)
        @JvmStatic fun <T> failed(exception: Throwable): Outcome<T> =
            Failed(exception)
        @JvmStatic fun <T> cancelled(): Outcome<T> =
            Cancelled()
    }
}
